sklearn/linear_model/coordinate_descent.py

```diff
@@ -696,28 +727,64 @@ def fit(self, X, y, check_input=True):
+
+        all_solvers = ['cd', 'saga']
+        if self.solver not in all_solvers:
+            raise ValueError("ElasticNet Regression supports only solvers in"
+                             " {}, got {}.".format(all_solvers, self.solver))
+
         if isinstance(self.precompute, str):
             raise ValueError('precompute should be one of True, False or'
                              ' array-like. Got %r' % self.precompute)
+        elif self.precompute is not False and self.solver == 'saga':
+            warnings.warn("The solver 'saga' does not use the precomputed "
+                          " Gram matrix", category=UserWarning)
+
+        if self.positive and self.solver == 'saga':
+            raise ValueError("ElasticNet supports positive contraints for "
+                             "coefficients only for solver 'cd', "
+                             "got solver={}, positive={}"
+                             .format(self.solver, self.positive))
 
         # Remember if X is copied
         X_copied = False
-        # We expect X and y to be float64 or float32 Fortran ordered arrays
-        # when bypassing checks
+        # For 'cd', we expect X and y to be float64 or float32 Fortran ordered
+        # arrays when bypassing checks
+        # For 'saga' we expect float64 C ordered arrays
         if check_input:
             X_copied = self.copy_X and self.fit_intercept
-            X, y = check_X_y(X, y, accept_sparse='csc',
-                             order='F', dtype=[np.float64, np.float32],
-                             copy=X_copied, multi_output=True, y_numeric=True)
-            y = check_array(y, order='F', copy=False, dtype=X.dtype.type,
-                            ensure_2d=False)
+            if self.solver == 'cd':
+                X, y = check_X_y(X, y, accept_sparse='csc',
+                                 order='F', dtype=[np.float64, np.float32],
+                                 copy=X_copied, multi_output=True,
+                                 y_numeric=True)
+                y = check_array(y, order='F', copy=False, dtype=X.dtype.type,
+                                ensure_2d=False)
+            else:
+                X, y = check_X_y(X, y, accept_sparse='csr',
+                                 order='C', dtype=np.float64,
+                                 copy=X_copied, multi_output=True,
+                                 y_numeric=True)
+                # For 2D target, we need contiguous columns as they are
+                # to be regressed independently on X
+                y = check_array(y, order='F', copy=False, dtype=X.dtype.type,
+                                ensure_2d=False)
 
         # Ensure copying happens only once, don't do it again if done above
         should_copy = self.copy_X and not X_copied
+        if self.solver == 'cd':
+            precompute_ = self.precompute
+        else:
+            precompute_ = False
```

Comment:
It may help if you just raise the warning here, at the same time when you assume `False` for `precompute`. It kinda make reading the code easier. Or set the `precompute_` up there when you raise the warning.

Reply:
I'will do so.